src/test/java/echoserver/EchoServerServiceTest.java

```diff
@@ -0,0 +1,130 @@
+package echoserver;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.ArrayList;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+
+public class EchoServerServiceTest {
+
+    private ArrayList<FakeClientConnection> clients = new ArrayList<FakeClientConnection>();
+    private EchoServerService testServer;
+    private ServerSocket hostSocket;
+    private FakeHost host;
+    private Logger echoLogger;
+
+    class FakeHost implements ConnectionHub {
+        ServerSocket hostSocket;
+
+        FakeHost(ServerSocket hostSocket) {
+            this.hostSocket = hostSocket;
+        }
+
+        public ConnectionDataStream listenForClientConnection()  {
+            return clients.remove(0);
+        }
+
+        public void close() throws IOException {
+            hostSocket.close();
+        }
+
+        public Boolean isClosed() {
+            return false;
+        }
+    }
+
+    class FakeClientConnection implements ConnectionDataStream {
+        FakeClientConnection(Socket socket) {
```

Comment:
Does `FakeClientConnection` need a `Socket`?

Reply:
Yes. When testing the entire run of the program, i mock '#readInput' so the program doesnt hang since that function directly reads the input from the socket. Typically, ths socket will halt the program and wait for the user to enter input.

Reply:
Gotcha. I guess what I am wondering about is whether the test double really needs a `Socket` in order to verify behavior.

Reply:
Its not verifying the behavior of the socket, it's just there so the program can run with real objects. For 'EchoServerService', i wanted to implicitly test that the host will stop the while loop after a null client is returned. I could have mocked it, but i wanted to try mimic the production environment.